### DISCUSSION TITLE
What is the architecture of AB Download Manager?

### DISCUSSION THREAD

**Discussion by AliYar-Khan** [amir1376/ab-download-manager#318] [Q&A] [answered]
I am building a cross platform download manager for all platforms including windows, Linux and macOS using flutter
Had some questions that anyone can answer, it will be very helpful

1️⃣ Should there be a client-server model?
Should downloads be handled directly by the app, or should a lightweight server process manage them for scalability and better resource management?

2️⃣ What database would you suggest?
For storing download history and metadata, would SQLite be sufficient, or should I explore alternatives like Hive, Drift, or Realm?

3️⃣ How to implement browser integration?
Should I rely on browser extensions, or are there simpler ways to capture and manage download links directly from browsers?

**Comment by amir1376** [comment, marked as answer]
Hi there.

You can find all of your question in the source code.

but If you I want to answer you shortly

> 1️⃣ Should there be a client-server model?

No. only client is required. 

> 2️⃣ What database would you suggest?

I personally use filesystem as my data source. but you can use database as well.
this can be easily changed in your data source implementation details

> How to implement browser integration?

It requires you to be familiar with `javascript` / `typescript` and  write plugins for browsers.
browsers give you an api (which is javascript) and you can use that api to capture downloads etc.

**Comment by cdlus** [comment]
I would answer it differently. 

The server/client model is a lot more efficient. 

For storing data SQLite is the best, for local storage. You can also then easily port it for a remote DB later, if you change how the data is stored. 

As for the browser extension, you want to look at the chrome.webRequest and chrome.downloads events, to capture a file request/download before they are complete. Also remember to send the cookies and headers to the server, so it can use the same data as the browser, which will reduce the chance of the requests failing by a lot.

**Reply by AliYar-Khan** [reply]
As chrome extension I want to have a download button whenever I watch a video on any social media. How that can be achieved ?

**Reply by cdlus** [reply]
> As chrome extension I want to have a download button whenever I watch a video on any social media. How that can be achieved ?

There are likely two techniques you can use to detect videos in a webpage. The simplest method would be using `document.querySelectorAll("video")` to get all the video tags in the webpage, and then loop over then adding a button that uses a relative position, so it appears in a certain location to the video.

The second and more advanced method is to get the link using chrome.webRequests and then checking the HTML/Dom for the same link.

I also imagine they have some custom search methods for popular sites that try and make it hard to do.

**Comment by dixtdf** [comment]
Hi! I recently built an independent open-source project that may be relevant to this discussion:

https://github.com/dixtdf/abdm-server

It is a self-hosted / headless Web download manager inspired by AB Download Manager, mainly intended for LAN, NAS and home-server use.

Some of the current features include:

* Docker deployment
* Web UI
* REST API + WebSocket
* download queue and concurrency control
* HTTP multi-connection downloading with 1–256 connections
* changing the connection count while a download is running
* pause/resume and restart recovery
* SQLite persistence
* English and Simplified Chinese UI
* amd64 / arm64 images

The project is independent from AB Download Manager and does not fork the upstream repository. AB Download Manager is tracked as a Git submodule, with an optional adapter layer for its download engine.

Sharing it here in case it is useful to anyone looking for a headless/server/Web-oriented ABDM workflow.
